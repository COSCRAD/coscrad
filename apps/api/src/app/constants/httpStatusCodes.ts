import { ValueType } from '../../lib/types/valueType';

const httpStatusCodes = {
    ok: 200,
    badRequest: 400,
    unauthorized: 401,
    forbidden: 403,
    notFound: 404,
    internalError: 500,
};

export default httpStatusCodes;

export type HttpStatusCode = ValueType<typeof httpStatusCodes>;
