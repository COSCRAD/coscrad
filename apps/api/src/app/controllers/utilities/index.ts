// TODO Barrel export everything in this directory and fix existing imports
export * from './mixLinkIntoViewModelDescription';
