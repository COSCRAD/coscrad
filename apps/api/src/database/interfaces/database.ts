import { EntityId } from '../../domain/types/entity-id';
import { Maybe } from '../../lib/types/maybe';

export interface IDatabase {
  fetchById: <TEntityDTO>(
    id: EntityId,
    collectionName: string
  ) => Promise<Maybe<TEntityDTO>>;

  fetchMany: <TEntityDTO>(
    collectionName: string
  ) => Promise<Maybe<TEntityDTO[]>>;

  create: <TEntityDTO>(
    dto: TEntityDTO,
    collectionName: string
  ) => Promise<void>;

  createMany: <TEntityDTO>(
    dto: TEntityDTO[],
    collectionName: string
  ) => Promise<void>;

  getCount: (collectionName: string) => Promise<number>;

  update: <TEntityDTO>(
    id: EntityId,
    dto: TEntityDTO,
    collectionName: string
  ) => Promise<void>;
}
