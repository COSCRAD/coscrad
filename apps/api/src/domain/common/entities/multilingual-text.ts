import {
    IMultilingualText,
    IMultlingualTextItem,
    LanguageCode,
    MultilingualTextItemRole,
} from '@coscrad/api-interfaces';
import {
    ExternalEnum,
    NestedDataType,
    NonEmptyString,
    TypeDecoratorOptions,
} from '@coscrad/data-types';
import { InternalError } from '../../../lib/errors/InternalError';
import { Maybe } from '../../../lib/types/maybe';
import { NotFound } from '../../../lib/types/not-found';
import { DTO } from '../../../types/DTO';
import { DeepPartial } from '../../../types/DeepPartial';
import { ResultOrError } from '../../../types/ResultOrError';
import { Valid } from '../../domainModelValidators/Valid';
import BaseDomainModel from '../../models/BaseDomainModel';
import { DuplicateLanguageInMultilingualTextError } from '../../models/audio-item/errors/duplicate-language-in-multilingual-text.error';
import { MultilingualTextHasNoOriginalError } from '../../models/audio-item/errors/multilingual-text-has-no-original.error';
import { MultipleOriginalsInMultilingualTextError } from '../../models/audio-item/errors/multiple-originals-in-multilingual-text.error';
import { isNull, isUndefined } from '../../utilities/validation/is-null-or-undefined';

export { MultilingualTextItemRole };

export const LanguageCodeEnum = (options: TypeDecoratorOptions) =>
    ExternalEnum(
        {
            enumLabel: `language code`,
            enumName: `LanguageCode`,
            labelsAndValues: Object.entries(LanguageCode).map(([label, languageCode]) => ({
                label,
                value: languageCode,
            })),
        },
        options
    );

export class MultilingualTextItem extends BaseDomainModel implements IMultlingualTextItem {
    @ExternalEnum(
        {
            labelsAndValues: Object.entries(LanguageCode).map(([label, value]) => ({
                label,
                value,
            })),
            enumLabel: 'Language_Code',
            enumName: 'LangaugeCode',
        },
        {
            label: 'language code',
            description: 'an official identifier of the language',
        }
    )
    readonly languageCode: LanguageCode;

    @NonEmptyString({
        label: 'text',
        description: 'plain text in the given language',
    })
    readonly text: string;

    @ExternalEnum(
        {
            labelsAndValues: Object.values(MultilingualTextItemRole).map((label) => ({
                value: label,
                label,
            })),
            enumLabel: 'text item role',
            enumName: 'Multilingual Text Item Role',
        },
        {
            description: 'role of this text in the translation process',
            label: 'text item role',
        }
    )
    readonly role: MultilingualTextItemRole;

    // TODO Support (user defined) dialects of the same language - would tags be the better way to do this?
    // @NonEmptyString({
    //     label: 'dialect',
    //     description: 'an informal identifier of the dialect of the given language',
    //     isOptional: true,
    // })
    // readonly dialect?: string;

    constructor(dto: DTO<MultilingualTextItem>) {
        super();

        if (!dto) return;

        const { languageCode, role, text } = dto;

        this.languageCode = languageCode;

        this.role = role;

        this.text = text;
    }
}

export class MultilingualText extends BaseDomainModel implements IMultilingualText {
    @NestedDataType(MultilingualTextItem, {
        label: 'items',
        description: 'one item for each provided language',
        isArray: true,
    })
    readonly items: MultilingualTextItem[];

    constructor(dto: DTO<MultilingualText>) {
        super();

        if (!dto) return;

        const { items } = dto;

        this.items = Array.isArray(items)
            ? items.map((item) => new MultilingualTextItem(item))
            : null;
    }

    toString(): string {
        /**
         * This is necessary when building error messages in the validation layer.
         */
        if (!Array.isArray(this.items)) {
            if (isNull(this.items)) return 'null';

            if (isUndefined(this.items)) return 'undefined';

            return 'invalid';
        }

        return this.items.map(({ text, languageCode }) => `{${languageCode}}: ${text}`).join('\n');
    }

    translate(languageCode: LanguageCode): Maybe<MultilingualTextItem> {
        return this.items.find((item) => item.languageCode === languageCode) || NotFound;
    }

    getOriginalTextItem(): MultilingualTextItem {
        return this.items.find(({ role }) => role === MultilingualTextItemRole.original).clone();
    }

    append(item: MultilingualTextItem): MultilingualText {
        return this.clone({
            // avoid shared references
            items: this.items.concat(item).map((item) => new MultilingualTextItem(item)),
        } as DeepPartial<DTO<this>>);
    }

    validateComplexInvariants(): ResultOrError<Valid> {
        const allErrors: InternalError[] = [];

        /**
         * We can't validate complex invariants if the simple invariants (data types)
         * are off. Maybe we should skip the call to `validateComplexInvariants`
         * in that case?
         */
        if (!Array.isArray(this.items)) return;

        const originalTextItems = this.items.filter(
            ({ role }) => role === MultilingualTextItemRole.original
        );

        if (originalTextItems.length > 1)
            allErrors.push(
                new MultipleOriginalsInMultilingualTextError(
                    originalTextItems.map(({ languageCode }) => languageCode)
                )
            );

        if (originalTextItems.length === 0)
            allErrors.push(new MultilingualTextHasNoOriginalError());

        // Initialize a map for counting the items with each language
        const initialItemCountMap = Object.values(LanguageCode).reduce(
            (accMap, languageCode) => accMap.set(languageCode, 0),
            new Map<LanguageCode, number>()
        );

        const countsForEachLanguage = this.items.reduce(
            (accMap, { languageCode }) => accMap.set(languageCode, accMap.get(languageCode) + 1),
            initialItemCountMap
        );

        const languageDuplicationErrors = [...countsForEachLanguage.entries()]
            .filter(([_key, count]) => count > 1)
            .map(([key, _count]) => key)
            .map((key: LanguageCode) => new DuplicateLanguageInMultilingualTextError(key));

        allErrors.push(...languageDuplicationErrors);

        return allErrors.length > 0
            ? new InternalError(`Encountered an instance of invalid multilingual text`, allErrors)
            : Valid;
    }
}
