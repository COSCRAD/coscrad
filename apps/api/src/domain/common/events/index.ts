export * from './build-event-handler-maps';
export * from './coscrad-event-factory';
export * from './coscrad-event-sourced-aggregate-factory';
export * from './coscrad-event-union';
export * from './coscrad-event.decorator';
export * from './event.module';
