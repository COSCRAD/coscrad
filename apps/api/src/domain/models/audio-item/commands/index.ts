export * from './constants';
export * from './create-audio-item';
export * from './transcripts';
