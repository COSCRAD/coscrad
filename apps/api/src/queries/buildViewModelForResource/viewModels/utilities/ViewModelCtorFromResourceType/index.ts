export * from './aggregateTypeToViewModelCtor';
export * from './types';
