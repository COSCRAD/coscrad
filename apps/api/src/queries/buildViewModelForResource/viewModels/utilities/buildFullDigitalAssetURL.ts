export default (baseURL: string, filename: string, extension = 'mp3'): string => {
    return `${baseURL}${filename}.${extension}`;
};
