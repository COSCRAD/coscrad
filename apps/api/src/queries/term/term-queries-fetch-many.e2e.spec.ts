import {
    AggregateType,
    CoscradUserRole,
    HttpStatusCode,
    LanguageCode,
    MultilingualTextItemRole,
} from '@coscrad/api-interfaces';
import { INestApplication } from '@nestjs/common';
import * as request from 'supertest';
import httpStatusCodes from '../../app/constants/httpStatusCodes';
import setUpIntegrationTest from '../../app/controllers/__tests__/setUpIntegrationTest';
import getValidAggregateInstanceForTest from '../../domain/__tests__/utilities/getValidAggregateInstanceForTest';
import { MultilingualText } from '../../domain/common/entities/multilingual-text';
import buildDummyUuid from '../../domain/models/__tests__/utilities/buildDummyUuid';
import { AccessControlList } from '../../domain/models/shared/access-control/access-control-list.entity';
import { TermCreated } from '../../domain/models/term/commands';
import {
    ITermQueryRepository,
    TERM_QUERY_REPOSITORY_TOKEN,
} from '../../domain/models/term/queries';
import { CoscradUserGroup } from '../../domain/models/user-management/group/entities/coscrad-user-group.entity';
import { CoscradUserWithGroups } from '../../domain/models/user-management/user/entities/user/coscrad-user-with-groups';
import { CoscradUser } from '../../domain/models/user-management/user/entities/user/coscrad-user.entity';
import { FullName } from '../../domain/models/user-management/user/entities/user/full-name.entity';
import { clonePlainObjectWithOverrides } from '../../lib/utilities/clonePlainObjectWithOverrides';
import { ArangoCollectionId } from '../../persistence/database/collection-references/ArangoCollectionId';
import { ArangoDatabaseProvider } from '../../persistence/database/database.provider';
import TestRepositoryProvider from '../../persistence/repositories/__tests__/TestRepositoryProvider';
import generateDatabaseNameForTestSuite from '../../persistence/repositories/__tests__/generateDatabaseNameForTestSuite';
import buildTestDataInFlatFormat from '../../test-data/buildTestDataInFlatFormat';
import { TestEventStream } from '../../test-data/events';
import { TermViewModel } from '../buildViewModelForResource/viewModels/term.view-model';

const indexEndpoint = `/resources/terms`;

// We require an existing user for ACL tests
const dummyQueryUserId = buildDummyUuid(4);

const { user: users, userGroup: userGroups } = buildTestDataInFlatFormat();

const dummyUser = (users as CoscradUser[])[0].clone({
    authProviderUserId: `auth0|${dummyQueryUserId}`,
    id: dummyQueryUserId,
    roles: [CoscradUserRole.viewer],
});

const dummyGroup = (userGroups as CoscradUserGroup[])[0].clone({ userIds: [dummyUser.id] });

const dummyUserWithGroups = new CoscradUserWithGroups(dummyUser, [dummyGroup]);

// Set up test data (use event sourcing to set up state)
const termText = `Term (in the language)`;

const originalLanguage = LanguageCode.Haida;

const termTranslation = `Term (in English)`;

const translationLanguage = LanguageCode.English;

const termId = buildDummyUuid(101);

const termCompositeIdentifier = {
    type: AggregateType.term,
    id: termId,
};

const termIdUnpublishedNoUserAccessId = buildDummyUuid(102);

const termIdUnpublishedWithUserAccessId = buildDummyUuid(103);

const dummyContributorFirstName = 'Dumb';

const dummyContributorLastName = 'McContributor';

const dummyContributor = getValidAggregateInstanceForTest(AggregateType.contributor).clone({
    fullName: new FullName({
        firstName: dummyContributorFirstName,
        lastName: dummyContributorLastName,
    }),
});

const termCreated = new TestEventStream().andThen<TermCreated>({
    type: 'TERM_CREATED',
    payload: {
        text: termText,
        languageCode: originalLanguage,
    },
    meta: {
        // we also want to check that the contributions come through
        contributorIds: [dummyContributor.id],
    },
});

const dummyTermView = clonePlainObjectWithOverrides(
    TermViewModel.fromTermCreated(termCreated.as(termCompositeIdentifier)[0] as TermCreated),
    {
        contributions: [
            {
                id: dummyContributor.id,
                fullName: dummyContributor.fullName.toString(),
            },
        ],
    }
);

const publicTermView = clonePlainObjectWithOverrides(dummyTermView, {
    isPublished: true,
    name: new MultilingualText(dummyTermView.name).translate({
        text: termTranslation,
        languageCode: translationLanguage,
        role: MultilingualTextItemRole.freeTranslation,
        // we are insisting by casting that the call to `translate` won't fail above
    }) as MultilingualText,
});

const privateTermThatUserCanAccess = clonePlainObjectWithOverrides(publicTermView, {
    id: termIdUnpublishedWithUserAccessId,
    isPublished: false,
    accessControlList: new AccessControlList().allowUser(dummyQueryUserId),
});

const privateTermUserCannotAccess = clonePlainObjectWithOverrides(publicTermView, {
    id: termIdUnpublishedNoUserAccessId,
    isPublished: false,
    // no special access
    accessControlList: new AccessControlList(),
});

// const promptTermId = buildDummyUuid(2)

describe(`when querying for a term: fetch many`, () => {
    const testDatabaseName = generateDatabaseNameForTestSuite();

    let app: INestApplication;

    let testRepositoryProvider: TestRepositoryProvider;

    let databaseProvider: ArangoDatabaseProvider;

    let termQueryRepository: ITermQueryRepository;

    let seedTerms: (terms: TermViewModel[]) => Promise<void>;

    afterAll(async () => {
        await app.close();

        databaseProvider.close();
    });

    describe(`when the user is unauthenticated`, () => {
        beforeAll(async () => {
            ({ app, testRepositoryProvider, databaseProvider } = await setUpIntegrationTest(
                {
                    ARANGO_DB_NAME: testDatabaseName,
                }
                // no authenticated user
            ));

            await app.init();

            termQueryRepository = app.get(TERM_QUERY_REPOSITORY_TOKEN);

            seedTerms = async (terms: TermViewModel[]) => {
                await termQueryRepository.createMany(terms);
            };
        });

        describe(`when there is a published term in the index view`, () => {
            beforeEach(async () => {
                await databaseProvider
                    .getDatabaseForCollection(ArangoCollectionId.contributors)
                    .clear();

                await databaseProvider.getDatabaseForCollection('term__VIEWS').clear();

                await testRepositoryProvider.getContributorRepository().create(dummyContributor);

                await seedTerms([
                    publicTermView,
                    privateTermThatUserCanAccess,
                    privateTermUserCannotAccess,
                ]);
            });

            it(`should only return published terms`, async () => {
                const res = await request(app.getHttpServer()).get(indexEndpoint);

                expect(res.status).toBe(httpStatusCodes.ok);

                const {
                    body: { entities },
                } = res;

                // TODO check that the actions are correct !!!!!
                expect(entities).toHaveLength(1);

                // Only one published result should come through from eventHistoryForMany
                const result = entities[0] as TermViewModel;

                expect(result.id).toBe(termId);

                const _knownContributors = await testRepositoryProvider
                    .getContributorRepository()
                    .fetchMany();

                expect(
                    result.contributions.some(({ fullName, id: foundContributorId }) => {
                        return (
                            // TODO we will want a test helper for this comparison in case we decide to change format
                            fullName ===
                                `${dummyContributorFirstName} ${dummyContributorLastName}` &&
                            foundContributorId === dummyContributor.id
                        );
                    })
                ).toBe(true);
            });
        });

        describe(`when there are no published terms`, () => {
            beforeEach(async () => {
                await databaseProvider
                    .getDatabaseForCollection(ArangoCollectionId.contributors)
                    .clear();

                await databaseProvider.getDatabaseForCollection('term__VIEWS').clear();

                await testRepositoryProvider.getContributorRepository().create(dummyContributor);

                /**
                 * Note that there is no ordinary system user authenticated for the request
                 * in this scenario. This is a public request.
                 */
                await seedTerms([privateTermUserCannotAccess, privateTermThatUserCanAccess]);
            });

            it(`should return no terms`, async () => {
                const res = await request(app.getHttpServer()).get(indexEndpoint);

                expect(res.status).toBe(httpStatusCodes.ok);

                expect(res.body.entities).toHaveLength(0);
            });
        });
    });

    describe(`when the user is authenticated as a non-admin user`, () => {
        beforeAll(async () => {
            ({ app, testRepositoryProvider, databaseProvider } = await setUpIntegrationTest(
                {
                    ARANGO_DB_NAME: testDatabaseName,
                },
                {
                    testUserWithGroups: dummyUserWithGroups,
                }
            ));

            termQueryRepository = app.get(TERM_QUERY_REPOSITORY_TOKEN);

            seedTerms = async (terms: TermViewModel[]) => {
                await termQueryRepository.createMany(terms);
            };
        });

        describe(`when there is a term that is unpublished`, () => {
            describe(`when the user does not have read access`, () => {
                beforeEach(async () => {
                    await databaseProvider.getDatabaseForCollection('term__VIEWS').clear();

                    await databaseProvider
                        .getDatabaseForCollection(ArangoCollectionId.contributors)
                        .clear();

                    await seedTerms([
                        publicTermView,
                        privateTermThatUserCanAccess,
                        // This one should not be visible to an ordinary user
                        privateTermUserCannotAccess,
                    ]);

                    await testRepositoryProvider
                        .getContributorRepository()
                        .create(dummyContributor);
                });

                it(`should not return the unpublished term`, async () => {
                    const res = await request(app.getHttpServer()).get(indexEndpoint);

                    expect(res.status).toBe(httpStatusCodes.ok);

                    /**
                     * + private, but user in ACL
                     * - private, user not in ACL
                     * + published
                     */
                    expect(res.body.entities).toHaveLength(2);
                });
            });

            describe(`when the user does have read access`, () => {
                beforeEach(async () => {
                    await databaseProvider.getDatabaseForCollection('term__VIEWS').clear();

                    await databaseProvider
                        .getDatabaseForCollection(ArangoCollectionId.contributors)
                        .clear();

                    await seedTerms([
                        // we only seed the accessible term here
                        privateTermThatUserCanAccess,
                    ]);

                    await testRepositoryProvider
                        .getContributorRepository()
                        .create(dummyContributor);
                });

                it(`should return the unpublished term`, async () => {
                    const res = await request(app.getHttpServer()).get(indexEndpoint);

                    expect(res.status).toBe(httpStatusCodes.ok);

                    const {
                        body: { entities },
                    } = res;

                    // this is the first and only term we seeded
                    const result = entities[0] as TermViewModel;

                    expect(result.id).toBe(termIdUnpublishedWithUserAccessId);
                });
            });
        });
    });

    (
        [
            [CoscradUserRole.superAdmin, 'when the user is a COSCRAD admin'],
            [CoscradUserRole.projectAdmin, 'when the user is a project admin'],
        ] as const
    ).forEach(([userRole, description]) => {
        describe(description, () => {
            beforeAll(async () => {
                ({ app, testRepositoryProvider, databaseProvider } = await setUpIntegrationTest(
                    {
                        ARANGO_DB_NAME: testDatabaseName,
                    },
                    {
                        testUserWithGroups: new CoscradUserWithGroups(
                            dummyUser.clone({
                                roles: [userRole],
                            }),
                            []
                        ),
                    }
                ));

                termQueryRepository = app.get(TERM_QUERY_REPOSITORY_TOKEN);

                seedTerms = async (terms: TermViewModel[]) => {
                    await termQueryRepository.createMany(terms);
                };
            });

            beforeEach(async () => {
                await databaseProvider.getDatabaseForCollection('term__VIEWS').clear();

                await databaseProvider
                    .getDatabaseForCollection(ArangoCollectionId.contributors)
                    .clear();

                await seedTerms([
                    publicTermView,
                    privateTermThatUserCanAccess,
                    privateTermUserCannotAccess,
                ]);

                await testRepositoryProvider.getContributorRepository().create(dummyContributor);
            });

            it(`should allow the user to access private resources`, async () => {
                const res = await request(app.getHttpServer()).get(indexEndpoint);

                const numberOfPrivateTerms = 2;

                const numberOfPublicTerms = 1;

                expect(res.status).toBe(HttpStatusCode.ok);

                expect(res.body.entities).toHaveLength(numberOfPrivateTerms + numberOfPublicTerms);

                /**
                 * This if statement is effectively a filter for our test
                 * case builder pattern. We only want to snapshot the response
                 * once for one specific class of user. We want this to be
                 * an admin user, because admin users have access to all
                 * data, including command info (available actions).
                 * Finally, we do this for the project admin, as queries
                 * for this user are by far more common than for coscrad admin
                 * in practice.
                 */
                if (userRole === CoscradUserRole.projectAdmin) {
                    /**
                     * This is to catch breaking changes in the API contract with
                     * the client. See the above comment for the corresponding detail
                     * endpoint test case.
                     */
                    expect(res.body).toMatchSnapshot();
                }
            });
        });
    });
});
