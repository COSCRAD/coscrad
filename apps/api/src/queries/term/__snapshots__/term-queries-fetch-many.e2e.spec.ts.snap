// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when querying for a term: fetch many when the user is a project admin should allow the user to access private resources 1`] = `
{
  "entities": [
    {
      "_id": "term__VIEWS/9b1deb4d-3b7d-4bad-9bdd-2b0d7b100102",
      "accessControlList": {
        "allowedGroupIds": [],
        "allowedUserIds": [],
      },
      "actions": [
        {
          "description": "Translate an existing term (typically from the language to English)",
          "form": {
            "fields": [
              {
                "constraints": [
                  {
                    "message": "must be a non-empty string",
                    "name": "non-empty string",
                  },
                  {
                    "message": "must be a defined value",
                    "name": "defined value",
                  },
                ],
                "description": "translation for the given term",
                "label": "translation",
                "name": "translation",
                "type": "TEXT_FIELD",
              },
              {
                "constraints": [
                  {
                    "message": "Must be a valid \${propertyLabel}",
                    "name": "IS_ENUM",
                  },
                  {
                    "message": "Required",
                    "name": "defined value",
                  },
                ],
                "description": "language in which you are translating the term (typically English)",
                "label": "language code",
                "name": "languageCode",
                "options": [
                  {
                    "display": "Chilcotin",
                    "value": "clc",
                  },
                  {
                    "display": "Haida",
                    "value": "hai",
                  },
                  {
                    "display": "English",
                    "value": "en",
                  },
                  {
                    "display": "French",
                    "value": "fra",
                  },
                  {
                    "display": "Chinook",
                    "value": "chn",
                  },
                  {
                    "display": "Zapotec",
                    "value": "zap",
                  },
                  {
                    "display": "Spanish",
                    "value": "spa",
                  },
                ],
                "type": "STATIC_SELECT",
              },
            ],
            "prepopulatedFields": {
              "aggregateCompositeIdentifier": {
                "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100102",
                "type": "term",
              },
            },
          },
          "label": "Translate Term",
          "type": "TRANSLATE_TERM",
        },
      ],
      "contributions": [],
      "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100102",
      "isPublished": false,
      "name": {
        "items": [
          {
            "languageCode": "hai",
            "role": "original",
            "text": "Term (in the language)",
          },
          {
            "languageCode": "en",
            "role": "free translation",
            "text": "Term (in English)",
          },
        ],
      },
    },
    {
      "_id": "term__VIEWS/9b1deb4d-3b7d-4bad-9bdd-2b0d7b100103",
      "accessControlList": {
        "allowedGroupIds": [],
        "allowedUserIds": [
          "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100004",
        ],
      },
      "actions": [
        {
          "description": "Translate an existing term (typically from the language to English)",
          "form": {
            "fields": [
              {
                "constraints": [
                  {
                    "message": "must be a non-empty string",
                    "name": "non-empty string",
                  },
                  {
                    "message": "must be a defined value",
                    "name": "defined value",
                  },
                ],
                "description": "translation for the given term",
                "label": "translation",
                "name": "translation",
                "type": "TEXT_FIELD",
              },
              {
                "constraints": [
                  {
                    "message": "Must be a valid \${propertyLabel}",
                    "name": "IS_ENUM",
                  },
                  {
                    "message": "Required",
                    "name": "defined value",
                  },
                ],
                "description": "language in which you are translating the term (typically English)",
                "label": "language code",
                "name": "languageCode",
                "options": [
                  {
                    "display": "Chilcotin",
                    "value": "clc",
                  },
                  {
                    "display": "Haida",
                    "value": "hai",
                  },
                  {
                    "display": "English",
                    "value": "en",
                  },
                  {
                    "display": "French",
                    "value": "fra",
                  },
                  {
                    "display": "Chinook",
                    "value": "chn",
                  },
                  {
                    "display": "Zapotec",
                    "value": "zap",
                  },
                  {
                    "display": "Spanish",
                    "value": "spa",
                  },
                ],
                "type": "STATIC_SELECT",
              },
            ],
            "prepopulatedFields": {
              "aggregateCompositeIdentifier": {
                "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100103",
                "type": "term",
              },
            },
          },
          "label": "Translate Term",
          "type": "TRANSLATE_TERM",
        },
      ],
      "contributions": [],
      "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100103",
      "isPublished": false,
      "name": {
        "items": [
          {
            "languageCode": "hai",
            "role": "original",
            "text": "Term (in the language)",
          },
          {
            "languageCode": "en",
            "role": "free translation",
            "text": "Term (in English)",
          },
        ],
      },
    },
    {
      "_id": "term__VIEWS/9b1deb4d-3b7d-4bad-9bdd-2b0d7b100101",
      "accessControlList": {
        "allowedGroupIds": [],
        "allowedUserIds": [],
      },
      "actions": [
        {
          "description": "Translate an existing term (typically from the language to English)",
          "form": {
            "fields": [
              {
                "constraints": [
                  {
                    "message": "must be a non-empty string",
                    "name": "non-empty string",
                  },
                  {
                    "message": "must be a defined value",
                    "name": "defined value",
                  },
                ],
                "description": "translation for the given term",
                "label": "translation",
                "name": "translation",
                "type": "TEXT_FIELD",
              },
              {
                "constraints": [
                  {
                    "message": "Must be a valid \${propertyLabel}",
                    "name": "IS_ENUM",
                  },
                  {
                    "message": "Required",
                    "name": "defined value",
                  },
                ],
                "description": "language in which you are translating the term (typically English)",
                "label": "language code",
                "name": "languageCode",
                "options": [
                  {
                    "display": "Chilcotin",
                    "value": "clc",
                  },
                  {
                    "display": "Haida",
                    "value": "hai",
                  },
                  {
                    "display": "English",
                    "value": "en",
                  },
                  {
                    "display": "French",
                    "value": "fra",
                  },
                  {
                    "display": "Chinook",
                    "value": "chn",
                  },
                  {
                    "display": "Zapotec",
                    "value": "zap",
                  },
                  {
                    "display": "Spanish",
                    "value": "spa",
                  },
                ],
                "type": "STATIC_SELECT",
              },
            ],
            "prepopulatedFields": {
              "aggregateCompositeIdentifier": {
                "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100101",
                "type": "term",
              },
            },
          },
          "label": "Translate Term",
          "type": "TRANSLATE_TERM",
        },
      ],
      "contributions": [],
      "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100101",
      "isPublished": true,
      "name": {
        "items": [
          {
            "languageCode": "hai",
            "role": "original",
            "text": "Term (in the language)",
          },
          {
            "languageCode": "en",
            "role": "free translation",
            "text": "Term (in English)",
          },
        ],
      },
    },
  ],
  "indexScopedActions": [
    {
      "description": "Create a new term in the language",
      "form": {
        "fields": [
          {
            "constraints": [
              {
                "message": "must be a non-empty string",
                "name": "non-empty string",
              },
              {
                "message": "must be a defined value",
                "name": "defined value",
              },
            ],
            "description": "text for the term (in the language)",
            "label": "text",
            "name": "text",
            "type": "TEXT_FIELD",
          },
          {
            "constraints": [
              {
                "message": "Must be a valid \${propertyLabel}",
                "name": "IS_ENUM",
              },
              {
                "message": "Required",
                "name": "defined value",
              },
            ],
            "description": "the language of this term",
            "label": "language code",
            "name": "languageCode",
            "options": [
              {
                "display": "Chilcotin",
                "value": "clc",
              },
              {
                "display": "Haida",
                "value": "hai",
              },
              {
                "display": "English",
                "value": "en",
              },
              {
                "display": "French",
                "value": "fra",
              },
              {
                "display": "Chinook",
                "value": "chn",
              },
              {
                "display": "Zapotec",
                "value": "zap",
              },
              {
                "display": "Spanish",
                "value": "spa",
              },
            ],
            "type": "STATIC_SELECT",
          },
        ],
        "prepopulatedFields": {},
      },
      "label": "Create Term",
      "type": "CREATE_TERM",
    },
    {
      "description": "Create a new Prompt Term in the language",
      "form": {
        "fields": [
          {
            "constraints": [
              {
                "message": "must be a non-empty string",
                "name": "non-empty string",
              },
              {
                "message": "must be a defined value",
                "name": "defined value",
              },
            ],
            "description": "text for the term (in English)",
            "label": "text",
            "name": "text",
            "type": "TEXT_FIELD",
          },
        ],
        "prepopulatedFields": {},
      },
      "label": "Create Prompt Term",
      "type": "CREATE_PROMPT_TERM",
    },
  ],
}
`;
